### ISSUE TITLE
The intermediate fusion method based on lift-splat-shoot got poor results on opv2v dataset

### ISSUE THREAD

**Issue by wangsh2001** [yifanlu0227/CoAlign#27]
Hi~
I've tested the intermediate fusion method(v2x-vit) and late fusion method based on lift-splat-shoot recently.However, v2x-vit got poor results 0.47/0.37/0.22 at ap30/50/70 and the late fusion baseline got 0.83/0.71/0.43 at ap30/50/70. Besides, no fusion baseline got 0.48/0.37/0.18 at ap30/50/70, which is similar to v2x-vit.

name: opv2v_lss_single_efficientnet
root_dir: "/data2/wsh/data/OPV2V/train"
validate_dir: "/data2/wsh/data/OPV2V/validate"
test_dir: "/data2/wsh/data/OPV2V/test"

yaml_parser: "load_lift_splat_shoot_params"
train_params:
  batch_size: &batch_size 2
  epoches: 50
  eval_freq: 2
  save_freq: 2
  max_cav: 7

input_source: ['camera']
label_type: 'camera'

comm_range: 70
only_vis_ego: true

add_data_extension: ['bev_visibility.png']

fusion:
  core_method: 'late'
  dataset: 'opv2v'
  args: 
    proj_first: false # useless
    grid_conf: &grid_conf
      xbound: [-51.2, 51.2, 0.4]   # 需要和preprocess一致. Need to be consistent with preprocess.
      ybound: [-51.2, 51.2, 0.4]   # 需要和preprocess一致. Need to be consistent with preprocess.
      zbound: [-10, 10, 20.0]   # 不需要和preprocess一致. NO Need to be consistent with preprocess.
      ddiscr: [2, 50, 48]
      mode: 'LID' # or 'UD'
    data_aug_conf: &data_aug_conf
      resize_lim: [0.8, 0.85]
      final_dim: [480, 640]
      rot_lim: [-3.6, 3.6]
      H: 600
      W: 800
      rand_flip: False
      bot_pct_lim: [0.0, 0.05]
      cams: ['camera0', 'camera1', 'camera2', 'camera3']
      Ncams: 4

# preprocess-related
preprocess:
  # options: BasePreprocessor, VoxelPreprocessor, BevPreprocessor
  core_method: 'SpVoxelPreprocessor'
  args:
    voxel_size: &voxel_size [0.4, 0.4, 4]
    max_points_per_voxel: 32
    max_voxel_train: 32000
    max_voxel_test: 70000
  # detection range for each individual cav.
  cav_lidar_range: &cav_lidar [-51.2, -51.2, -3, 51.2, 51.2, 1]

data_augment: # use less
  - NAME: random_world_flip
    ALONG_AXIS_LIST: [ 'x' ]

  - NAME: random_world_rotation
    WORLD_ROT_ANGLE: [ -0.78539816, 0.78539816 ]

  - NAME: random_world_scaling
    WORLD_SCALE_RANGE: [ 0.95, 1.05 ]

# anchor box related
postprocess:
  core_method: 'VoxelPostprocessor' # That's ok
  gt_range: *cav_lidar
  anchor_args:
    cav_lidar_range: *cav_lidar
    l: 3.9
    w: 1.6
    h: 1.56
    feature_stride: 2
    r: &anchor_yaw [0, 90]
    num: &achor_num 2
  target_args:
    pos_threshold: 0.6
    neg_threshold: 0.45
    score_threshold: 0.25
  order: 'hwl' # hwl or lwh
  max_num: 100 # maximum number of objects in a single frame. use this number to make sure different frames has the same dimension in the same batch
  nms_thresh: 0.15
  dir_args: &dir_args
    dir_offset: 0.7853
    num_bins: 2
    anchor_yaw: *anchor_yaw

# model related
model:
  core_method: lift_splat_shoot
  args:
    anchor_number: *achor_num
    grid_conf: *grid_conf
    data_aug_conf: *data_aug_conf
    dir_args: *dir_args
    img_downsample: 8
    img_features: 128
    use_depth_gt: false
    depth_supervision: false
    bevout_feature: 128

    shrink_header:
      kernal_size: [ 3 ]
      stride: [ 2 ]
      padding: [ 1 ]
      dim: [ 128 ]
      input_dim: 128
    camera_encoder: EfficientNet

loss:
  core_method: point_pillar_loss
  args:
    pos_cls_weight: 2.0
    cls:
      type: 'SigmoidFocalLoss'
      alpha: 0.25
      gamma: 2.0
      weight: 1.0
    reg:
      type: 'WeightedSmoothL1Loss'
      sigma: 3.0
      codewise: true
      weight: 2.0
    dir:
      type: 'WeightedSoftmaxClassificationLoss'
      weight: 0.2
      args: *dir_args

optimizer:
  core_method: Adam
  lr: 0.0015
  args:
    eps: 1e-10
    weight_decay: 1e-4

lr_scheduler:
  core_method: multistep #step, multistep and Exponential support
  gamma: 0.1
  step_size: [25, 40]

name: opv2v_lss_v2xvit
root_dir: "/data1/wsh/data/OPV2V/train"
validate_dir: "/data1/wsh/data/OPV2V/validate"
test_dir: "/data1/wsh/data/OPV2V/test"

yaml_parser: "load_lift_splat_shoot_params"     # we need specific loading functions for different backbones.
train_params:          # the common training parameters
  batch_size: &batch_size 2
  epoches: 50
  eval_freq: 2
  save_freq: 2
  max_cav: 7

input_source: ['camera']          # 'lidar', 'camera', 'depth'
label_type: 'camera'              # 'lidar' or 'camera'

comm_range: 70                    
only_vis_ego: true

add_data_extension: ['bev_visibility.png']

fusion:
  core_method: 'intermediate'
  dataset: 'opv2v'
  args: 
    proj_first: false             # useless
    grid_conf: &grid_conf
      xbound: [-51.2, 51.2, 0.4]  # 需要和preprocess一致. Need to be consistent with preprocess.
      ybound: [-51.2, 51.2, 0.4]  # 需要和preprocess一致. Need to be consistent with preprocess.
      zbound: [-10, 10, 20.0]     # 不需要和preprocess一致. No Need to be consistent with preprocess.
      ddiscr: [2, 50, 48]         # depth_min, depth_max, num_bins, make grid in image plane
      mode: 'LID'
    data_aug_conf: &data_aug_conf
      resize_lim: [0.8, 0.85]
      final_dim: [480, 640]
      rot_lim: [-3.6, 3.6]
      H: 600
      W: 800
      rand_flip: False
      bot_pct_lim: [0.0, 0.05]
      cams: ['camera0', 'camera1', 'camera2', 'camera3']
      Ncams: 4

# preprocess-related
preprocess:
  # options: BasePreprocessor, VoxelPreprocessor, BevPreprocessor
  core_method: 'SpVoxelPreprocessor'
  args:
    voxel_size: &voxel_size [0.4, 0.4, 4]         # the voxel resolution
    max_points_per_voxel: 32                      # maximum points allowed in each voxel
    max_voxel_train: 32000                        # the maximum voxel number during training
    max_voxel_test: 70000                         # the maximum voxel number during testing
  # detection range for each individual cav.
  cav_lidar_range: &cav_lidar [-51.2, -51.2, -3, 51.2, 51.2, 1]

data_augment:                     # useless
  - NAME: random_world_flip
    ALONG_AXIS_LIST: [ 'x' ]

  - NAME: random_world_rotation
    WORLD_ROT_ANGLE: [ -0.78539816, 0.78539816 ]

  - NAME: random_world_scaling
    WORLD_SCALE_RANGE: [ 0.95, 1.05 ]

# anchor box related
postprocess:
  core_method: 'VoxelPostprocessor' # VoxelPostprocessor, BevPostprocessor supported
  gt_range: *cav_lidar
  anchor_args:                  # anchor generator parameters
    cav_lidar_range: *cav_lidar # the range is consistent with the lidar cropping range to generate the correct anchors
    l: 3.9                      # the default length of the anchor
    w: 1.6                      # the default width
    h: 1.56                     # the default height
    feature_stride: 2           # the feature map is shrank twice compared the input voxel tensor
    r: &anchor_yaw [0, 90]      # the yaw angles. 0, 90 meaning for each voxel, two anchors will be generated with 0 and 90 degree yaw angle
    num: &achor_num 2           # for each location in the feature map, 2 anchors will be generated
  target_args:                  # used to generate positive and negative samples for object detection
    pos_threshold: 0.6
    neg_threshold: 0.45
    score_threshold: 0.25
  order: 'hwl' # hwl or lwh
  max_num: 100 # maximum number of objects in a single frame. use this number to make sure different frames has the same dimension in the same batch
  nms_thresh: 0.15
  dir_args: &dir_args
    dir_offset: 0.7853    # pi / 4
    num_bins: 2
    anchor_yaw: *anchor_yaw

# model related
model:
  core_method: lift_splat_shoot_intermediate  # trainer will load the corresponding model python file with the same name
  args:  # detailed parameters of this model
    anchor_number: *achor_num
    grid_conf: *grid_conf
    data_aug_conf: *data_aug_conf
    dir_args: *dir_args
    img_downsample: 8
    img_features: &img_feature 128
    use_depth_gt: false
    depth_supervision: false
    supervise_single: true                            
    bevout_feature: 128
    camera_encoder: EfficientNet

    fusion_args:
      core_method: v2xvit
      args:
        voxel_size: *voxel_size
        in_channels: *img_feature
        v2xvit:
          transformer:
            encoder: &encoder
              # number of fusion blocks per encoder layer, V2XFusionBlock
              num_blocks: 1
              # number of encoder layers, BaseWindowAttention
              depth: 3
              use_roi_mask: true
              use_RTE: &use_RTE false
              RTE_ratio: &RTE_ratio 0 # 2 means the dt has 100ms interval while 1 means 50 ms interval
              # agent-wise attention, HMSA(Heterogeneous multi-agent self-attention)
              cav_att_config: &cav_att_config
                dim: 256
                use_hetero: true
                use_RTE: *use_RTE
                RTE_ratio: *RTE_ratio
                heads: 8
                dim_head: 32
                dropout: 0.3
              # spatial-wise attention, BaseWindowAttention
              pwindow_att_config: &pwindow_att_config
                dim: 256
                heads: [16, 8, 4]
                dim_head: [16, 32, 64]
                dropout: 0.3
                window_size: [4, 8, 16]
                relative_pos_embedding: true
                fusion_method: 'split_attn'
              # feedforward condition
              feed_forward: &feed_forward
                mlp_dim: 256
                dropout: 0.3
              sttf: &sttf
                voxel_size: *voxel_size
                downsample_rate: 2

loss:         # loss function
  core_method: point_pillar_loss        # trainer will load the loss function with the same name
  args:
    pos_cls_weight: 2.0
    cls:
      type: 'SigmoidFocalLoss'
      alpha: 0.25
      gamma: 2.0
      weight: 1.0                       # classification weights
    reg:
      type: 'WeightedSmoothL1Loss'
      sigma: 3.0
      codewise: true
      weight: 2.0                       # regression weights

optimizer:                              # optimzer setup
  core_method: Adam                     # the name has to exist in Pytorch optimizer library
  lr: 0.0015
  args:
    eps: 1e-10
    weight_decay: 1e-4

lr_scheduler:                           # learning rate schedular
  core_method: multistep                # step, multistep and Exponential support
  gamma: 0.1
  step_size: [25, 40]

**Comment by TillBeemelmanns** [comment]
Hi @wangsh0111 Could you resolve this problem ?
